### ISSUE TITLE
[CLI][yolov8]: Wandb Callback ultralytics CUDA tensor error

### ISSUE THREAD

**Issue by robertomancebom** [wandb/wandb#7004]
### Bug description

Hi there 👋
I am working with YOLOv8 from Ultralytics with wandb. I tried using the add_wandb_callback function to keep track of the output artifacts (I want to ignore model checkpoints) while finetuning a pretrained YOLOv8m, but I got an error telling me about incompatible tensors (input with CUDA and model weights ). This is my code:

```python

import wandb
from wandb.integration.ultralytics import add_wandb_callback
from ultralytics import YOLO

DATA_PATH = "/dataset/path"

run = wandb.init(entity="e", project="p", 
                 job_type="j",
                 group='g',
                 name="n",
                 config=configs)
                 
# Load pretrained model
model = YOLO("yolov8m.pt")

# Wandb Callback
add_wandb_callback(model, enable_model_checkpointing=False)

# Train model
results = model.train(data=DATA_PATH + "data.yaml", 
            epochs=configs['EPOCHS'], 
            patience=configs['PATIENCE'],
            imgsz=configs['IMG_SIZE'],
            batch=configs['BATCH_SIZE'],
            optimizer=configs['OPTIMIZER'],
            lr0=configs['INIT_LR'],
            project=project_dir,
            name=yolo_name
           )
```
And this is the error I get:

```shell
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
Cell In[8], line 8
      5 add_wandb_callback(model, enable_model_checkpointing=False)
      7 # Train model
----> 8 results = model.train(data=DATA_PATH + "data.yaml", 
      9             epochs=configs['EPOCHS'], 
     10             patience=configs['PATIENCE'],
     11             imgsz=configs['IMG_SIZE'],
     12             batch=configs['BATCH_SIZE'],
     13             optimizer=configs['OPTIMIZER'],
     14             lr0=configs['INIT_LR'],
     15             project=project_dir,
     16             name=yolo_name
     17            )

File /opt/conda/lib/python3.10/site-packages/ultralytics/engine/model.py:601, in Model.train(self, trainer, **kwargs)
    598             pass
    600 self.trainer.hub_session = self.session  # attach optional HUB session
--> 601 self.trainer.train()
    602 # Update model and cfg after training
    603 if RANK in (-1, 0):

File /opt/conda/lib/python3.10/site-packages/ultralytics/engine/trainer.py:208, in BaseTrainer.train(self)
    205         ddp_cleanup(self, str(file))
    207 else:
--> 208     self._do_train(world_size)

File /opt/conda/lib/python3.10/site-packages/ultralytics/engine/trainer.py:424, in BaseTrainer._do_train(self, world_size)
    422 # Validation
    423 if self.args.val or final_epoch or self.stopper.possible_stop or self.stop:
--> 424     self.metrics, self.fitness = self.validate()
    425 self.save_metrics(metrics={**self.label_loss_items(self.tloss), **self.metrics, **self.lr})
    426 self.stop |= self.stopper(epoch + 1, self.fitness) or final_epoch

File /opt/conda/lib/python3.10/site-packages/ultralytics/engine/trainer.py:544, in BaseTrainer.validate(self)
    538 def validate(self):
    539     """
    540     Runs validation on test set using self.validator.
    541 
    542     The returned dict is expected to contain "fitness" key.
    543     """
--> 544     metrics = self.validator(self)
    545     fitness = metrics.pop("fitness", -self.loss.detach().cpu().numpy())  # use loss as fitness measure if not found
    546     if not self.best_fitness or self.best_fitness < fitness:

File /opt/conda/lib/python3.10/site-packages/torch/utils/_contextlib.py:115, in context_decorator.<locals>.decorate_context(*args, **kwargs)
    112 @functools.wraps(func)
    113 def decorate_context(*args, **kwargs):
    114     with ctx_factory():
--> 115         return func(*args, **kwargs)

File /opt/conda/lib/python3.10/site-packages/ultralytics/engine/validator.py:198, in BaseValidator.__call__(self, trainer, model)
    196 self.finalize_metrics()
    197 self.print_results()
--> 198 self.run_callbacks("on_val_end")
    199 if self.training:
    200     model.float()

File /opt/conda/lib/python3.10/site-packages/ultralytics/engine/validator.py:266, in BaseValidator.run_callbacks(self, event)
    264 """Runs all callbacks associated with a specified event."""
    265 for callback in self.callbacks.get(event, []):
--> 266     callback(self)

File /opt/conda/lib/python3.10/site-packages/torch/utils/_contextlib.py:115, in context_decorator.<locals>.decorate_context(*args, **kwargs)
    112 @functools.wraps(func)
    113 def decorate_context(*args, **kwargs):
    114     with ctx_factory():
--> 115         return func(*args, **kwargs)

File /opt/conda/lib/python3.10/site-packages/wandb/integration/ultralytics/callback.py:363, in WandBUltralyticsCallback.on_val_end(self, trainer)
    354     self.validation_table = plot_segmentation_validation_results(
    355         dataloader=dataloader,
    356         class_label_map=class_label_map,
   (...)
    360         max_validation_batches=self.max_validation_batches,
    361     )
    362 elif self.task == "detect":
--> 363     self.validation_table = plot_detection_validation_results(
    364         dataloader=dataloader,
    365         class_label_map=class_label_map,
    366         model_name=self.model_name,
    367         predictor=self.predictor,
    368         table=self.validation_table,
    369         max_validation_batches=self.max_validation_batches,
    370     )
    371 elif self.task == "classify":
    372     self.validation_table = plot_classification_validation_results(
    373         dataloader=dataloader,
    374         model_name=self.model_name,
   (...)
    377         max_validation_batches=self.max_validation_batches,
    378     )

File /opt/conda/lib/python3.10/site-packages/wandb/integration/ultralytics/bbox_utils.py:165, in plot_detection_validation_results(dataloader, class_label_map, model_name, predictor, table, max_validation_batches, epoch)
    163 max_validation_batches = min(max_validation_batches, num_dataloader_batches)
    164 for batch_idx, batch in enumerate(dataloader):
--> 165     prediction_results = predictor(batch["im_file"])
    166     progress_bar_result_iterable = tqdm(
    167         enumerate(prediction_results),
    168         total=len(prediction_results),
    169         desc=f"Generating Visualizations for batch-{batch_idx + 1}/{max_validation_batches}",
    170     )
    171     for img_idx, prediction_result in progress_bar_result_iterable:

File /opt/conda/lib/python3.10/site-packages/ultralytics/engine/predictor.py:204, in BasePredictor.__call__(self, source, model, stream, *args, **kwargs)
    202     return self.stream_inference(source, model, *args, **kwargs)
    203 else:
--> 204     return list(self.stream_inference(source, model, *args, **kwargs))

File /opt/conda/lib/python3.10/site-packages/torch/utils/_contextlib.py:35, in _wrap_generator.<locals>.generator_context(*args, **kwargs)
     32 try:
     33     # Issuing `None` to a generator fires it up
     34     with ctx_factory():
---> 35         response = gen.send(None)
     37     while True:
     38         try:
     39             # Forward the response to our caller and get its next request

File /opt/conda/lib/python3.10/site-packages/ultralytics/engine/predictor.py:283, in BasePredictor.stream_inference(self, source, model, *args, **kwargs)
    281 # Inference
    282 with profilers[1]:
--> 283     preds = self.inference(im, *args, **kwargs)
    284     if self.args.embed:
    285         yield from [preds] if isinstance(preds, torch.Tensor) else preds  # yield embedding tensors

File /opt/conda/lib/python3.10/site-packages/ultralytics/engine/predictor.py:140, in BasePredictor.inference(self, im, *args, **kwargs)
    134 """Runs inference on a given image using the specified model and arguments."""
    135 visualize = (
    136     increment_path(self.save_dir / Path(self.batch[0][0]).stem, mkdir=True)
    137     if self.args.visualize and (not self.source_type.tensor)
    138     else False
    139 )
--> 140 return self.model(im, augment=self.args.augment, visualize=visualize, embed=self.args.embed, *args, **kwargs)

File /opt/conda/lib/python3.10/site-packages/torch/nn/modules/module.py:1518, in Module._wrapped_call_impl(self, *args, **kwargs)
   1516     return self._compiled_call_impl(*args, **kwargs)  # type: ignore[misc]
   1517 else:
-> 1518     return self._call_impl(*args, **kwargs)

File /opt/conda/lib/python3.10/site-packages/torch/nn/modules/module.py:1527, in Module._call_impl(self, *args, **kwargs)
   1522 # If we don't have any hooks, we want to skip the rest of the logic in
   1523 # this function, and just call forward.
   1524 if not (self._backward_hooks or self._backward_pre_hooks or self._forward_hooks or self._forward_pre_hooks
   1525         or _global_backward_pre_hooks or _global_backward_hooks
   1526         or _global_forward_hooks or _global_forward_pre_hooks):
-> 1527     return forward_call(*args, **kwargs)
   1529 try:
   1530     result = None

File /opt/conda/lib/python3.10/site-packages/ultralytics/nn/autobackend.py:384, in AutoBackend.forward(self, im, augment, visualize, embed)
    381     im = im.permute(0, 2, 3, 1)  # torch BCHW to numpy BHWC shape(1,320,192,3)
    383 if self.pt or self.nn_module:  # PyTorch
--> 384     y = self.model(im, augment=augment, visualize=visualize, embed=embed)
    385 elif self.jit:  # TorchScript
    386     y = self.model(im)

File /opt/conda/lib/python3.10/site-packages/torch/nn/modules/module.py:1518, in Module._wrapped_call_impl(self, *args, **kwargs)
   1516     return self._compiled_call_impl(*args, **kwargs)  # type: ignore[misc]
   1517 else:
-> 1518     return self._call_impl(*args, **kwargs)

File /opt/conda/lib/python3.10/site-packages/torch/nn/modules/module.py:1527, in Module._call_impl(self, *args, **kwargs)
   1522 # If we don't have any hooks, we want to skip the rest of the logic in
   1523 # this function, and just call forward.
   1524 if not (self._backward_hooks or self._backward_pre_hooks or self._forward_hooks or self._forward_pre_hooks
   1525         or _global_backward_pre_hooks or _global_backward_hooks
   1526         or _global_forward_hooks or _global_forward_pre_hooks):
-> 1527     return forward_call(*args, **kwargs)
   1529 try:
   1530     result = None

File /opt/conda/lib/python3.10/site-packages/ultralytics/nn/tasks.py:83, in BaseModel.forward(self, x, *args, **kwargs)
     81 if isinstance(x, dict):  # for cases of training and validating while training.
     82     return self.loss(x, *args, **kwargs)
---> 83 return self.predict(x, *args, **kwargs)

File /opt/conda/lib/python3.10/site-packages/ultralytics/nn/tasks.py:101, in BaseModel.predict(self, x, profile, visualize, augment, embed)
     99 if augment:
    100     return self._predict_augment(x)
--> 101 return self._predict_once(x, profile, visualize, embed)

File /opt/conda/lib/python3.10/site-packages/ultralytics/nn/tasks.py:122, in BaseModel._predict_once(self, x, profile, visualize, embed)
    120 if profile:
    121     self._profile_one_layer(m, x, dt)
--> 122 x = m(x)  # run
    123 y.append(x if m.i in self.save else None)  # save output
    124 if visualize:

File /opt/conda/lib/python3.10/site-packages/torch/nn/modules/module.py:1518, in Module._wrapped_call_impl(self, *args, **kwargs)
   1516     return self._compiled_call_impl(*args, **kwargs)  # type: ignore[misc]
   1517 else:
-> 1518     return self._call_impl(*args, **kwargs)

File /opt/conda/lib/python3.10/site-packages/torch/nn/modules/module.py:1527, in Module._call_impl(self, *args, **kwargs)
   1522 # If we don't have any hooks, we want to skip the rest of the logic in
   1523 # this function, and just call forward.
   1524 if not (self._backward_hooks or self._backward_pre_hooks or self._forward_hooks or self._forward_pre_hooks
   1525         or _global_backward_pre_hooks or _global_backward_hooks
   1526         or _global_forward_hooks or _global_forward_pre_hooks):
-> 1527     return forward_call(*args, **kwargs)
   1529 try:
   1530     result = None

File /opt/conda/lib/python3.10/site-packages/ultralytics/nn/modules/conv.py:54, in Conv.forward_fuse(self, x)
     52 def forward_fuse(self, x):
     53     """Perform transposed convolution of 2D data."""
---> 54     return self.act(self.conv(x))

File /opt/conda/lib/python3.10/site-packages/torch/nn/modules/module.py:1518, in Module._wrapped_call_impl(self, *args, **kwargs)
   1516     return self._compiled_call_impl(*args, **kwargs)  # type: ignore[misc]
   1517 else:
-> 1518     return self._call_impl(*args, **kwargs)

File /opt/conda/lib/python3.10/site-packages/torch/nn/modules/module.py:1527, in Module._call_impl(self, *args, **kwargs)
   1522 # If we don't have any hooks, we want to skip the rest of the logic in
   1523 # this function, and just call forward.
   1524 if not (self._backward_hooks or self._backward_pre_hooks or self._forward_hooks or self._forward_pre_hooks
   1525         or _global_backward_pre_hooks or _global_backward_hooks
   1526         or _global_forward_hooks or _global_forward_pre_hooks):
-> 1527     return forward_call(*args, **kwargs)
   1529 try:
   1530     result = None

File /opt/conda/lib/python3.10/site-packages/torch/nn/modules/conv.py:460, in Conv2d.forward(self, input)
    459 def forward(self, input: Tensor) -> Tensor:
--> 460     return self._conv_forward(input, self.weight, self.bias)

File /opt/conda/lib/python3.10/site-packages/torch/nn/modules/conv.py:456, in Conv2d._conv_forward(self, input, weight, bias)
    452 if self.padding_mode != 'zeros':
    453     return F.conv2d(F.pad(input, self._reversed_padding_repeated_twice, mode=self.padding_mode),
    454                     weight, bias, self.stride,
    455                     _pair(0), self.dilation, self.groups)
--> 456 return F.conv2d(input, weight, bias, self.stride,
    457                 self.padding, self.dilation, self.groups)

RuntimeError: Input type (torch.cuda.FloatTensor) and weight type (torch.FloatTensor) should be the same
```

### Environment

WandB version: 0.16.3

OS: Ubuntu 20.04.6 LTS

Python version: Python 3.10.12

Versions of relevant libraries:
ultralytics==8.1.14

### Additional Context

Note: I have already tried using model.to('cuda') after loading the pretrained model but the error is the same.

**Comment by thanos-wandb** [comment]
Hi @robertomancebom thank you for reporting this issue and for providing us with a code example to reproduce it. Let me please surface this issue with our Growth/Integrations team and we will follow-up here with you.

**Comment by gauritomar** [comment]
Hey @robertomancebom, 
The error you're encountering (RuntimeError: Input type (torch.cuda.FloatTensor) and weight type (torch.FloatTensor) should be the same) typically occurs when there's a mismatch between the types of tensors (input and weights) when running inference on your YOLOv8 model.

To resolve this issue, you need to ensure that both the model and the input tensors are on the same device, cuda gpu in your case.

```
# Move the model to GPU if available
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
model = model.to(device)
```

**Comment by Firestl** [comment]
same here

**Comment by soumik12345** [comment]
Hi @robertomancebom, the advanced features for the Ultralytics integration enabled by `add_wandb_callback` were developed and tested with `ultralyticsv8.0.238` and below. You can still use the integration without the advanced features by simply having `wandb` installed.

**Comment by robertomancebom** [comment]
@soumik12345 I tested with the `ultralytics `version you mentioned `8.0.238` but the same issue still happens. It is curious that it happens in the second epoch, the first one does not raise any errors.

**Comment by saikrishna-prathapaneni** [comment]
> @soumik12345 I tested with the `ultralytics `version you mentioned `8.0.238` but the same issue still happens. It is curious that it happens in the second epoch, the first one does not raise any errors.

same goes with me

**Comment by manfredmichael** [comment]
I have a technical assessment so I can't wait for the next release.

However, I downgraded both ultralytics & wandb to make it work on Google Colab

```
pip install wandb==0.15.12
pip install ultralytics==8.0.186
```

**Comment by thanos-wandb** [comment]
Hi @manfredmichael @saikrishna-prathapaneni @robertomancebom @gauritomar we're closing this ticket since the issue has been addressed in our March release v0.16.4. Please let us know if you're still experiencing this issue after `pip install --upgrade wandb`, and we'll be happy to keep investigating.
